### PR TITLE
Cli display graph

### PR DESCRIPTION
**Pull request by burnout87**

### CODE REVIEW

**Comment by burnout87** [comment]
A first output

graph

**Comment by volodymyrss** [comment]
add provenance graph visualization, as already shown by the renku

**Comment by volodymyrss** [comment]
following today's discussion, we want to replicate and then extend https://renkulab.io/projects/vladimir.savchenko/oda-sdss/files/lineage/final-an.ipynb

**Comment by burnout87** [comment]
As discussed with @andriineronov , here is a first implementation of a more clear and clean output graph

graph

**Comment by burnout87** [comment]
I managed to build the graph containing the information regarding the execution, meaning:

- input
- parameters
- outputs
It is basically the same graph produced by the lineage on Renku

graph_no_da_info

**Comment by volodymyrss** [comment]
nice! what about colors and so? Also I would put the inputs in separate nodes, see as in https://github.com/oda-hub/smartsky/issues/15#issuecomment-852118331 they are blue. Not too critical, but ok.

**Comment by burnout87** [comment]
> nice! what about colors and so? Also I would put the inputs in separate nodes, see as in [oda-hub/smartsky#15 (comment)](https://github.com/oda-hub/smartsky/issues/15#issuecomment-852118331) they are blue. Not too critical, but ok.

It's all about playing with the (extremely powerful) functionalities provided by the rdflib, that enables to navigate the graph, and modify, also graphically if we want.

I will play around a bit with that and see.

**Comment by volodymyrss** [comment]
Basically I think it would be nice it looked concise, like this https://github.com/oda-hub/smartsky/issues/15#issuecomment-852118331
Or at least how https://renkulab.io/projects/vladimir.savchenko/oda-sdss/files/lineage/final-an.ipynb uses colors and shapes to represent some attributes.
But we can discuss how useful it is. It may seem decorative, but effective visualization aids thinking, which is why it's important.

**Comment by burnout87** [comment]
Improved version

graph

**Comment by burnout87** [comment]
> 
> Cool! I would further separate `command parameter` content into block with title `name_input` and value `NC 1068`

Do you mean have a dedicated node per parameter? So, in this specific case, to have one node with the title `name_input` and inside content `NGC 1068`?

**Comment by volodymyrss** [comment]
> > Oh I remember we also talked that Arrow between Action and AstroqueryModule could be reversed. It's because logically astroquery module is used in the action. Similarly to input parameters. It should be, as is now, different color than inputs, since it is a different kind of input-like thing.
> 
> Thanks! I remember that part. So something like this?

yes. pretty close. Maybe purely for presentation would be nice to have filled colors and no gray hats. and use node shape to represent some characteristics (e.g. instead of changing colors of astroquerymodule/source, change them to other shape) See renku graph and andrri graph. 

image


> 
> image
> 
> Regarding the color, should we change the one for the AstroqueryModule or the AstrophysicalObject ?

**Comment by burnout87** [comment]
Yes all the infos are there. Now I have to investigate on how to produce different shapes.

**Comment by volodymyrss** [comment]
nice! Some colors a slightly harsh. Why not use all the same colors as renku? Also "commandinput" not useful, enough just the notebook name, like renku. 
But in Action, unlike renku, I'd put just "papermill". `ns1:command` not too useful.

**Comment by burnout87** [comment]
Made some colors adjustment to make it resemble more the renku lineage one

image

**Comment by burnout87** [comment]
And a first implementation capturing the `get_images`

image

**Comment by volodymyrss** [comment]
By the way, this is very suspicious:

image

I guess it's RA and Dec. If yes, they should be marked. But if yes, it's impossible, since Dec can not be more than 90. Maybe it's not ordered?

**Comment by burnout87** [comment]
> By the way, this is very suspicious:
> 
> image
> 
> I guess it's RA and Dec. If yes, they should be marked. But if yes, it's impossible, since Dec can not be more than 90. Maybe it's not ordered?

Ok I will mark them.

**Comment by burnout87** [comment]
> 
> I guess it's RA and Dec. If yes, they should be marked. But if yes, it's impossible, since Dec can not be more than 90. Maybe it's not ordered?

I just double-checked and it is DEC and RA.

**Comment by volodymyrss** [comment]
> > I guess it's RA and Dec. If yes, they should be marked. But if yes, it's impossible, since Dec can not be more than 90. Maybe it's not ordered?
> 
> I just double-checked and it is DEC and RA.

Ok so please switch them. And maybe mark too.

**Comment by burnout87** [comment]
> > > I guess it's RA and Dec. If yes, they should be marked. But if yes, it's impossible, since Dec can not be more than 90. Maybe it's not ordered?
> > 
> > 
> > I just double-checked and it is DEC and RA.
> 
> Ok so please switch them. And maybe mark too.

ok, any suggestions also the Angle?

**Comment by burnout87** [comment]
Better this way?

image

**Comment by volodymyrss** [comment]
> Better this way?
> 
> image

Please do switch them. Nobody writes Dec, RA, only RA, Dec. 
For the angle, it's fine, but could be also written as "3.0 arcmin".

**Comment by burnout87** [comment]
> 
> Please do switch them. Nobody writes Dec, RA, only RA, Dec. For the angle, it's fine, but could be also written as "3.0 arcmin".

I guess also for the coordinates the `unit=` can be removed.

image

**Comment by volodymyrss** [comment]
> > Please do switch them. Nobody writes Dec, RA, only RA, Dec. For the angle, it's fine, but could be also written as "3.0 arcmin".
> 
> I guess also for the coordinates the `unit=` can be removed.
> 
> image

it's a bit unclear of deg applies to both or not.  But ok

**Comment by burnout87** [comment]
> 
> it's a bit unclear of deg applies to both or not. But ok

Ok I will put it to both, I don't know what the best and most used convention for its representation is.

**Comment by volodymyrss** [comment]
> > it's a bit unclear of deg applies to both or not. But ok
> 
> Ok I will put it to both, I just don't know what the best and most used convention for its representation is.

there are several ways. Putting for both is fine. But because it's a bit verbose common to write "222.0 22." always in ra, dec order.
Maybe let's use  the long form to avoid confusion. Deg on both.

**Comment by burnout87** [comment]
I should also change also that format for the coordinated used for the `get_images` with the SkyView, since not a SkyCoordinates was used.

image

**Comment by volodymyrss** [comment]
q

> I should also change also that format for the coordinated used for the `get_images` with the SkyView, since not a SkyCoordinates was used, just more general coordinates.
> 
> image

Hm, I am quite confused  about what it shows now. J2000 is epoch, related to coordinate frame. It's not coordinates. Does this SkyCoordinates instance have meaningful coordinates?

**Comment by burnout87** [comment]
> Hm, I am quite confused about what it shows now. J2000 is epoch, related to coordinate frame. It's not coordinates. Does this SkyCoordinates instance have meaningful coordinates?

No indeed, as I said, I think a different type of coordinates has been used and not SkyCoordinates, that's why it should be a bit more generic.

**Comment by volodymyrss** [comment]
> > Hm, I am quite confused about what it shows now. J2000 is epoch, related to coordinate frame. It's not coordinates. Does this SkyCoordinates instance have meaningful coordinates?
> 
> No indeed, as I said, I think a different type of coordinates has been used and not SkyCoordinates, that's why it should be a bit more generic.

If by "different type of coordinates" you mean not coordinates at all, then I understand and agree. There it should be a bit more generic in the sense that it should be completely different. E.g. Epoch (which J2000 is, as I said)

**Comment by burnout87** [comment]
> 
> If by "different type of coordinates" you mean not coordinates at all, then I understand and agree. There it should be a bit more generic in the sense that it should be completely different. E.g. Epoch (which J2000 is, as I said)

Why should they not be coordinates? If they are passed to the `get_images` call as such?

**Comment by burnout87** [comment]
You can see it here:

[https://github.com/oda-hub/oda-sdss/blob/master/final-an.ipynb](https://github.com/oda-hub/oda-sdss/blob/master/final-an.ipynb)

**Comment by volodymyrss** [comment]
> > If by "different type of coordinates" you mean not coordinates at all, then I understand and agree. There it should be a bit more generic in the sense that it should be completely different. E.g. Epoch (which J2000 is, as I said)
> 
> Why should they not be coordinates? If they are passed to the `get_images` call as such?

I did not yet check what is passed, what is shown is not coordinates.

**Comment by volodymyrss** [comment]
> > > If by "different type of coordinates" you mean not coordinates at all, then I understand and agree. There it should be a bit more generic in the sense that it should be completely different. E.g. Epoch (which J2000 is, as I said)
> > 
> > 
> > Why should they not be coordinates? If they are passed to the `get_images` call as such?
> 
> I did not yet check what is passed, what is shown is not coordinates.

Maybe there there are coordinates somewhere embedded along. But Epoch is different. And it's not generic: all SkyCoordinates must have a set of annotations (or, at least implicit assumptions), epoch is just one of them.

**Comment by burnout87** [comment]
> > > > If by "different type of coordinates" you mean not coordinates at all, then I understand and agree. There it should be a bit more generic in the sense that it should be completely different. E.g. Epoch (which J2000 is, as I said)
> > > 
> > > 
> > > Why should they not be coordinates? If they are passed to the `get_images` call as such?
> > 
> > 
> > I did not yet check what is passed, what is shown is not coordinates.
> 
> Maybe there there are coordinates somewhere embedded along. But Epoch is different. And it's not generic: all SkyCoordinates must have a set of annotations (or, at least implicit assumptions), epoch is just one of them.

Ok, then where can the issue be? In the notebook? in that specifc `get_images` implementation?

**Comment by volodymyrss** [comment]
> You can see it here:
> 
> https://github.com/oda-hub/oda-sdss/blob/master/final-an.ipynb

This is indeed, clearly, not coordinates. It's an short name for coordinate system. People sometimes use it as so in conversations, if they are not careful. They should be more careful. We have to be more careful, since we are building/using an ontology. If this is standard pattern of astroquery's get_images (to be verified) we should accept it, and derive true coordinates from `position`,  using `coordinates` keyword to clarify coordinate system. 

We should absolutely not propagate confusion, since we are doing the opposite and making an ontology.

**Comment by burnout87** [comment]
Just read this in the `get_images` documentation from astroquery:

```
coordinates : str
    Choose among common equatorial, galactic and ecliptic
    coordinate systems (``"J2000"``, ``"B1950"``, ``"Galactic"``,
    ``"E2000"``, ``"ICRS"``) or pass a custom string.
```

**Comment by volodymyrss** [comment]
> Just read this in the `get_images` documentation from astroquery:
> 
> ```
> coordinates : str
>     Choose among common equatorial, galactic and ecliptic
>     coordinate systems (``"J2000"``, ``"B1950"``, ``"Galactic"``,
>     ``"E2000"``, ``"ICRS"``) or pass a custom string.
> ```
> 
> But apparently it is not the case for query_region as far as I can tell.

interface of query region might not have it. But an assumption is implicitly there. It's `J2000`, we can almost safely assume.

**Comment by rokroskar** [comment]
This is looking really promising! It would be nice to maybe generalize this and include it in the main renku cli as a complement to the current [`renku workflow visualize`](https://renku.readthedocs.io/projects/renku-python/en/latest/commands.html#visualizing-executions). cc @panaetius

**Comment by volodymyrss** [comment]
we should refer to astroquery API as the reference. https://github.com/astropy/astroquery/blob/main/docs/api.rst . And deduce additional common assumptions when suitable, but with care.

**Comment by burnout87** [comment]
> we should refer to astroquery API as the reference. https://github.com/astropy/astroquery/blob/main/docs/api.rst . And deduce additional common assumptions when suitable, but with care.

Seems like all three methods make the same assumption regarding the interpretation of `coordinates`, but I could not find a unique definition for it, therefore I guess we have to adhere to the ones we find in the code comments/documentation.

**Comment by burnout87** [comment]
Here it's the output of the graph after a run of the Legacysurvey_tap.ipynb notbook
It's still using the old style of annotations, but is this what is expected?


image

**Comment by volodymyrss** [comment]
It looks ok I think. It'd be maybe best to converge on this branch of development.

**Comment by burnout87** [comment]
> It looks ok I think. It'd be maybe best to converge on this branch of development.

I just wanted to have some final feedback if it's ok as an output, before starting to optimize even further the annotation process

**Comment by volodymyrss** [comment]
> > It looks ok I think. It'd be maybe best to converge on this branch of development.
> 
> I just wanted to have some final feedback if it's ok as an output, before starting to optimize even further the annotation process

How output should be described depends on what this is going to be used with. This part is still a demonstrator: so it's enough. What's annotated makes sense.

**Comment by volodymyrss** [comment]
Could you add the summary, short tutorial, with pictures, into a `REAME.md`? Or another PR? Then let's make an issue.

**Comment by burnout87** [comment]
> Could you add the summary, short tutorial, with pictures, into a `REAME.md`? Or another PR? Then let's make an issue.

No I will make it here

**Comment by burnout87** [comment]
> Could you add the summary, short tutorial, with pictures, into a `REAME.md`? Or another PR? Then let's make an issue.
> 
> edit: like I did here https://github.com/oda-hub/renku-aqs/tree/upgraph

I wrote a brief summary with some code examples and pics, do you think is sufficient for now?

**Comment by burnout87** [comment]
I adapted with the new version of `renku-python`, and the output is still the one expected.

But the new way of handling the `workflows` some more investigation of the query on the graph is needed.

**Comment by burnout87** [comment]
> I adapted with the new version of `renku-python`, and the output is still the one expected.
> 
> But the new way of handling the `workflows` some more investigation of the query on the graph is needed.

Included also the case of a re-run of a `Workflow`

**Comment by burnout87** [comment]
Two different directions of development for the enhancement of the graph functionalities:

- pure javascript: based on using rdfjs library ([https://rdf.js.org/](https://rdf.js.org/), [https://github.com/rdfjs/N3.js](https://github.com/rdfjs/N3.js)):
  - lots of development, in a different language, of functionalities already developed in python
  - development of a separate web application
- flask endopoint that based on the clicked node information, returns the content of the graph to display
  - easier, 
  - setup of an endpoint, needs to explore options on ways to have it always available

**Comment by burnout87** [comment]
> Two different directions of development for the enhancement of the graph functionalities:
> 
>     * pure javascript: based on using rdfjs library (https://rdf.js.org/, https://github.com/rdfjs/N3.js):
>       
>       * lots of development, in a different language, of functionalities already developed in python
>       * development of a separate web application
> 
>     * flask endopoint that based on the clicked node information, returns the content of the graph to display
>       
>       * easier,
>       * setup of an endpoint, needs to explore options on ways to have it always available
> 
> 
> In both cases, it is no longer gonna be a simple cli call

Agreed on exploring a javascript-based approach

**Comment by volodymyrss** [comment]
> > Two different directions of development for the enhancement of the graph functionalities:
> > ```
> > * pure javascript: based on using rdfjs library (https://rdf.js.org/, https://github.com/rdfjs/N3.js):
> >   
> >   * lots of development, in a different language, of functionalities already developed in python
> >   * development of a separate web application
> > 
> > * flask endopoint that based on the clicked node information, returns the content of the graph to display
> >   
> >   * easier,
> >   * setup of an endpoint, needs to explore options on ways to have it always available
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In both cases, it is no longer gonna be a simple cli call
> 
> Agreed on exploring a javascript-based approach

Specifically, the js approach could be quite generic:  clicking on node `A` fetches from a (configurable) sparql list of nodes linked to the node `A`. This would be useful for many people.

**Comment by burnout87** [comment]
Functionality `renku aqs show-graph`  implemented

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,5 @@
 -e git+https://github.com/oda-hub/astroquery.git#egg=astroquery
--e git+https://github.com/oda-hub/aqsmodel-converters@simple-autolog#egg=aqsconverters
+-e git+https://github.com/oda-hub/aqsmodel-converters#egg=aqsconverters
```
Comment (volodymyrss) [inline requirements.txt:2]: we should constrain version of this, and also of renku-python

**Comment by burnout87** [comment]
Adaptions to the latest version of renku-python have been applied, I managed to run the `show-graph` command

**Comment by volodymyrss** [comment]
this happens to me after renku run:

```python
inject.ConstructorTypeError: <function _run_command at 0x7f5782728dc0> raised an error: <function Activity.from_plan at 0x7f579d5c1430> raised an error: renku_aqs_path() missing 1 required positional argument: 'client_dispatcher'

```

**Comment by volodymyrss** [comment]
by the way, to keep track of versions we can also test it in renkulab, we can insert it in https://github.com/oda-hub/contributed-project-templates.

**Comment by Panaetius** [comment]
> ```python
> renku_aqs_path
> ```

I think you need to do

```
from renku.command.command_builder import inject
[...]
@inject.autoparams("client_dispatcher")
def renku_aqs_path(...):
```
for the dependency injection. 
But we're (Finally!) getting rid of LocalClient in renku for a more sane solution, so a lot of this will change. Maybe we should have a small call to tell you about the differences. It's the last in a long line of refactorings to make the codebase cleaner.

**Comment by burnout87** [comment]
> 



> ```python
> ```python
> renku_aqs_path
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> 
> I think you need to do
> from renku.command.command_builder import inject
> [...]
> @inject.autoparams("client_dispatcher")
> def renku_aqs_path(...):
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> for the dependency injection.
> ```

thanks a lot @Panaetius ! I had actually just realized the dependency I was missing...

And I am looking forward to trying the newer version ! Yes let's plan a call specific for this, it's a good idea.

**Comment by volodymyrss** [comment]
> > 
> 
> > ```python
> > ```python
> > renku_aqs_path
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think you need to do
> > from renku.command.command_builder import inject
> > [...]
> > @inject.autoparams("client_dispatcher")
> > def renku_aqs_path(...):
> > for the dependency injection.
> 
> thanks a lot @Panaetius ! I had actually just realized the dependency I was missing...
> 
> And I am looking forward to trying the newer version ! Yes let's plan a call specific for this, it's a good idea.

It does not crash anymore! But it does not show anything, somehow selection does not yield anything.
If I loosen the selection (should it be an option to select more?), it produces larger graph but does not show it.
I am trying on https://renkulab.io/gitlab/astronomy/mmoda/fermi
We can see tomorrow with @burnout87, I'll be in Ecogia.

**Comment by burnout87** [comment]
> 
> It does not crash anymore! But it does not show anything, somehow selection does not yield anything. 

I am also not getting anything

> I am trying on https://renkulab.io/gitlab/astronomy/mmoda/fermi We can see tomorrow with @burnout87, I'll be in Ecogia.

Cool!

**Comment by burnout87** [comment]
[Screencast 2022-09-26 14:03:24.webm](https://user-images.githubusercontent.com/3757833/192271643-69e3f806-a831-4462-b54d-2549107d1575.webm)

**Comment by burnout87** [comment]
All the functionalities to handle the drawing and interaction with the graph are now imported from a dedicated javascript module.

This is maintained, for now in the repo https://github.com/burnout87/renku-aqs-graph-library , the repository is for now cloned and the module imported. Do we have a location where we can store such js-library and reference from there?

**Comment by volodymyrss** [comment]
Ok, cool!
I am not sure the name should be anything aqs-related, but I suppose it should be moved to oda-hub, and once you add a gh-pages branch, it will appear in odahub.io . Do you have the rights to move the repository?

It will also need a readme.

Meanwhile, you may be able to import it just like this:

https://raw.githubusercontent.com/burnout87/renku-aqs-graph-library/df5a7654567804a9d15a5ee8d03ec6636abe1d70/graph_helper.js

Maybe add simple example in readme with that?

**Comment by burnout87** [comment]
> Ok, cool! I am not sure the name should be anything aqs-related, but I suppose it should be moved to oda-hub, and once you add a gh-pages branch, it will appear in odahub.io . Do you have the rights to move the repository?

I should since I already did it once
> 
> It will also need a readme.
> 
> Meanwhile, you may be able to import it just like this:
> 
> https://raw.githubusercontent.com/burnout87/renku-aqs-graph-library/df5a7654567804a9d15a5ee8d03ec6636abe1d70/graph_helper.js

I will try that thanks

**Comment by burnout87** [comment]
> 
> https://raw.githubusercontent.com/burnout87/renku-aqs-graph-library/df5a7654567804a9d15a5ee8d03ec6636abe1d70/graph_helper.js

This link won't work for our purposes, we have to use services like rawGit (though for development at the moment, then a proper CDN would be the best solution) https://github.com/rgrove/rawgit/blob/master/FAQ.md

**Comment by volodymyrss** [comment]
> > https://raw.githubusercontent.com/burnout87/renku-aqs-graph-library/df5a7654567804a9d15a5ee8d03ec6636abe1d70/graph_helper.js
> 
> This link won't work for our purposes, we have to use services like rawGit (though for development at the moment, then a proper CDN would be the best solution) https://github.com/rgrove/rawgit/blob/master/FAQ.md

This was just an option before repository is transferred. Now we can just make a gh-pages branch: https://github.com/oda-hub/renku-aqs-graph-library/tree/gh-pages becomes https://odahub.io/renku-aqs-graph-library/ and https://odahub.io/renku-aqs-graph-library/utilities/graph_helper.js

**Comment by volodymyrss** [comment]
it's very weird we still did not merge it, it went to big. Let's merge and avoid this in the future

**Review by volodymyrss — APPROVED** [review]

**Pull request merged**